invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalConsumableUsageInArrear.java

```diff
@@ -248,9 +246,8 @@ protected UsageInArrearAggregate getToBeBilledUsageDetails(final LocalDate start
         return toBeBilledDetails;
     }
 
-    UsageConsumableInArrearTierUnitAggregate computeToBeBilledConsumableInArrearWith_TOP_TIER(final List<TieredBlock> tieredBlocks, final List<UsageConsumableInArrearTierUnitAggregate> previousUsage, final Long units) throws CatalogApiException {
-
-        Long remainingUnits = units;
+    UsageConsumableInArrearTierUnitAggregate computeToBeBilledConsumableInArrearWith_TOP_TIER(final List<TieredBlock> tieredBlocks, final List<UsageConsumableInArrearTierUnitAggregate> previousUsage, final BigDecimal units) throws CatalogApiException {
+        Long remainingUnits = units.longValue();
```

Comment:
We don't want any of these `longValue` conversion, the code should only manipulate `BigDecimal`, so I think something like that (untested):

```
           // If we don't change our catalog api, if not we can get rid of this conversion as well
            final BigDecimal tierBlockMax = new BigDecimal(tieredBlock.getMax());
            final BigDecimal tierBlockSize = new BigDecimal(tieredBlock.getSize());

            final BigDecimal[] divRem =  remainingUnits.divideAndRemainder(tierBlockSize);
            final BigDecimal tmp =  (divRem[1].compareTo(BigDecimal.ZERO) == 0) ? divRem[0] : divRem[0].add(BigDecimal.ONE);
            if (tmp.compareTo(tierBlockMax) > 0) {
                remainingUnits = remainingUnits.subtract(tierBlockMax.multiply(tierBlockSize));
            } else {
                targetBlock = tieredBlock;
                targetTierNum = tierNum;
                break;
            }

```